File: src/common/asset_registry.py
"""
에셋 레지스트리 관리 모듈
게임 내 모든 비주얼 에셋(3D 모델, 이미지, 애니메이션)의 경로를 관리합니다.
"""
import json
import os
from typing import Dict, Optional, Any


class AssetRegistry:
    """에셋 레지스트리 관리 클래스"""
    
    def __init__(self, registry_path: str = "data/assets_registry.json"):
        """
        Args:
            registry_path: 에셋 레지스트리 JSON 파일 경로
        """
        self.registry_path = registry_path
        self.registry: Dict[str, Any] = {}
        self._load_registry()
    
    def _load_registry(self):
        """레지스트리 파일 로드"""
        if os.path.exists(self.registry_path):
            with open(self.registry_path, 'r', encoding='utf-8') as f:
                self.registry = json.load(f)
        else:
            print(f"Warning: Asset registry not found at {self.registry_path}")
            self.registry = {}
    
    def get_champion_assets(self, champion_key: str) -> Optional[Dict[str, Any]]:
        """
        챔피언의 에셋 정보 조회
        
        Args:
            champion_key: 챔피언 키 (예: "Garen")
            
        Returns:
            에셋 정보 딕셔너리 또는 None
        """
        champions = self.registry.get("champions", {})
        return champions.get(champion_key)
    
    def get_building_assets(self, building_type: str, level: int = 1) -> Optional[Dict[str, Any]]:
        """
        건물의 에셋 정보 조회
        
        Args:
            building_type: 건물 타입 (예: "MAIN_CASTLE")
            level: 건물 레벨
            
        Returns:
            에셋 정보 딕셔너리 (레벨에 맞게 경로 치환됨)
        """
        buildings = self.registry.get("buildings", {})
        building_data = buildings.get(building_type)
        
        if building_data:
            # 레벨에 따라 경로 치환
            assets = building_data.get("assets", {}).copy()
            for key, path in assets.items():
                if isinstance(path, str) and "{level}" in path:
                    assets[key] = path.format(level=level)
            
            return {
                **building_data,
                "assets": assets
            }
        
        return None
    
    def get_tile_assets(self, resource_type: str, level: int = 1) -> Optional[Dict[str, Any]]:
        """
        타일의 에셋 정보 조회
        
        Args:
            resource_type: 자원 타입 (예: "FOOD", "WOOD")
            level: 타일 레벨
            
        Returns:
            에셋 정보 딕셔너리 (레벨에 맞게 경로 치환됨)
        """
        tiles = self.registry.get("tiles", {})
        tile_data = tiles.get(resource_type)
        
        if tile_data:
            # 레벨에 따라 경로 치환
            assets = tile_data.get("assets", {}).copy()
            for key, path in assets.items():
                if isinstance(path, str) and "{level}" in path:
                    assets[key] = path.format(level=level)
            
            return {
                **tile_data,
                "assets": assets
            }
        
        return None
    
    def get_effect_assets(self, effect_name: str) -> Optional[Dict[str, Any]]:
        """
        이펙트의 에셋 정보 조회
        
        Args:
            effect_name: 이펙트 이름 (예: "damage_hit")
            
        Returns:
            이펙트 에셋 정보
        """
        effects = self.registry.get("effects", {})
        return effects.get(effect_name)
    
    def get_ui_assets(self, category: str, name: str) -> Optional[str]:
        """
        UI 에셋 경로 조회
        
        Args:
            category: UI 카테고리 (예: "backgrounds", "buttons")
            name: 에셋 이름
            
        Returns:
            에셋 경로 문자열
        """
        ui = self.registry.get("ui", {})
        category_assets = ui.get(category, {})
        return category_assets.get(name)
    
    def get_fallback_color(self, entity_type: str, entity_key: str) -> str:
        """
        에셋이 없을 때 사용할 대체 색상 조회
        
        Args:
            entity_type: 엔티티 타입 ("champions", "buildings", "tiles")
            entity_key: 엔티티 키
            
        Returns:
            색상 코드 (예: "#4169E1")
        """
        entities = self.registry.get(entity_type, {})
        entity_data = entities.get(entity_key, {})
        fallback = entity_data.get("fallback", {})
        return fallback.get("color", "#FFFFFF")
    
    def asset_exists(self, asset_path: str, base_dir: str = "assets") -> bool:
        """
        에셋 파일이 실제로 존재하는지 확인
        
        Args:
            asset_path: 에셋 상대 경로
            base_dir: 기본 디렉토리
            
        Returns:
            파일 존재 여부
        """
        full_path = os.path.join(base_dir, asset_path)
        return os.path.exists(full_path)


# 전역 에셋 레지스트리 인스턴스
_asset_registry: Optional[AssetRegistry] = None


def get_asset_registry() -> AssetRegistry:
    """전역 에셋 레지스트리 인스턴스 반환"""
    global _asset_registry
    if _asset_registry is None:
        _asset_registry = AssetRegistry()
    return _asset_registry

File: flutter_app/lib/asset_registry.dart
import 'dart:convert';
import 'package:flutter/material.dart';
import 'package:flutter/services.dart';

/// 에셋 레지스트리 관리 클래스
/// 게임 내 모든 비주얼 에셋의 경로를 관리하고 로드합니다.
class AssetRegistry {
  static AssetRegistry? _instance;
  Map<String, dynamic> _registry = {};
  bool _isLoaded = false;

  AssetRegistry._();

  /// 싱글톤 인스턴스 반환
  static AssetRegistry get instance {
    _instance ??= AssetRegistry._();
    return _instance!;
  }

  /// 레지스트리 초기화 (앱 시작 시 호출)
  Future<void> initialize() async {
    if (_isLoaded) return;

    try {
      // 백엔드에서 에셋 레지스트리를 가져올 수도 있지만,
      // 현재는 로컬에 복사본을 두는 방식으로 구현
      // final String jsonString = await rootBundle.loadString('assets/assets_registry.json');
      // _registry = json.decode(jsonString);

      // 임시로 빈 레지스트리로 시작
      _registry = {};
      _isLoaded = true;

      debugPrint('Asset Registry initialized');
    } catch (e) {
      debugPrint('Failed to load asset registry: $e');
      _registry = {};
    }
  }

  /// 챔피언 에셋 정보 조회
  ChampionAssets? getChampionAssets(String championKey) {
    final champions = _registry['champions'] as Map<String, dynamic>?;
    if (champions == null) return null;

    final championData = champions[championKey] as Map<String, dynamic>?;
    if (championData == null) return null;

    return ChampionAssets.fromJson(championData);
  }

  /// 건물 에셋 정보 조회
  BuildingAssets? getBuildingAssets(String buildingType, int level) {
    final buildings = _registry['buildings'] as Map<String, dynamic>?;
    if (buildings == null) return null;

    final buildingData = buildings[buildingType] as Map<String, dynamic>?;
    if (buildingData == null) return null;

    return BuildingAssets.fromJson(buildingData, level);
  }

  /// 타일 에셋 정보 조회
  TileAssets? getTileAssets(String resourceType, int level) {
    final tiles = _registry['tiles'] as Map<String, dynamic>?;
    if (tiles == null) return null;

    final tileData = tiles[resourceType] as Map<String, dynamic>?;
    if (tileData == null) return null;

    return TileAssets.fromJson(tileData, level);
  }

  /// Fallback 색상 조회
  Color getFallbackColor(String entityType, String entityKey) {
    final entities = _registry[entityType] as Map<String, dynamic>?;
    if (entities == null) return Colors.grey;

    final entityData = entities[entityKey] as Map<String, dynamic>?;
    if (entityData == null) return Colors.grey;

    final fallback = entityData['fallback'] as Map<String, dynamic>?;
    if (fallback == null) return Colors.grey;

    final colorString = fallback['color'] as String?;
    if (colorString == null) return Colors.grey;

    return _parseColor(colorString);
  }

  Color _parseColor(String hexColor) {
    final hex = hexColor.replaceAll('#', '');
    return Color(int.parse('FF$hex', radix: 16));
  }
}

/// 챔피언 에셋 정보
class ChampionAssets {
  final String id;
  final String name;
  final String? model3d;
  final String? texture;
  final String? icon;
  final String? portrait;
  final Map<String, String> animations;
  final Color fallbackColor;
  final String fallbackText;

  ChampionAssets({
    required this.id,
    required this.name,
    this.model3d,
    this.texture,
    this.icon,
    this.portrait,
    this.animations = const {},
    required this.fallbackColor,
    required this.fallbackText,
  });

  factory ChampionAssets.fromJson(Map<String, dynamic> json) {
    final assets = json['assets'] as Map<String, dynamic>?;
    final fallback = json['fallback'] as Map<String, dynamic>?;
    final animations = assets?['animations'] as Map<String, dynamic>?;

    return ChampionAssets(
      id: json['id'] as String,
      name: json['name'] as String,
      model3d: assets?['model_3d'] as String?,
      texture: assets?['texture'] as String?,
      icon: assets?['icon'] as String?,
      portrait: assets?['portrait'] as String?,
      animations: animations?.map((k, v) => MapEntry(k, v as String)) ?? {},
      fallbackColor: _parseColor(fallback?['color'] as String? ?? '#FFFFFF'),
      fallbackText: fallback?['icon_text'] as String? ?? '?',
    );
  }

  static Color _parseColor(String hexColor) {
    final hex = hexColor.replaceAll('#', '');
    return Color(int.parse('FF$hex', radix: 16));
  }
}

/// 건물 에셋 정보
class BuildingAssets {
  final String id;
  final String name;
  final String? model3d;
  final String? texture;
  final String? icon;
  final Color fallbackColor;

  BuildingAssets({
    required this.id,
    required this.name,
    this.model3d,
    this.texture,
    this.icon,
    required this.fallbackColor,
  });

  factory BuildingAssets.fromJson(Map<String, dynamic> json, int level) {
    final assets = json['assets'] as Map<String, dynamic>?;
    final fallback = json['fallback'] as Map<String, dynamic>?;

    // 레벨에 따라 경로 치환
    String? model3d = assets?['model_3d'] as String?;
    if (model3d != null && model3d.contains('{level}')) {
      model3d = model3d.replaceAll('{level}', level.toString());
    }

    return BuildingAssets(
      id: json['id'] as String,
      name: json['name'] as String,
      model3d: model3d,
      texture: assets?['texture'] as String?,
      icon: assets?['icon'] as String?,
      fallbackColor: _parseColor(fallback?['color'] as String? ?? '#FFFFFF'),
    );
  }

  static Color _parseColor(String hexColor) {
    final hex = hexColor.replaceAll('#', '');
    return Color(int.parse('FF$hex', radix: 16));
  }
}

/// 타일 에셋 정보
class TileAssets {
  final String id;
  final String name;
  final String? model3d;
  final String? texture;
  final String? icon;
  final Color fallbackColor;

  TileAssets({
    required this.id,
    required this.name,
    this.model3d,
    this.texture,
    this.icon,
    required this.fallbackColor,
  });

  factory TileAssets.fromJson(Map<String, dynamic> json, int level) {
    final assets = json['assets'] as Map<String, dynamic>?;
    final fallback = json['fallback'] as Map<String, dynamic>?;

    // 레벨에 따라 경로 치환
    String? model3d = assets?['model_3d'] as String?;
    if (model3d != null && model3d.contains('{level}')) {
      model3d = model3d.replaceAll('{level}', level.toString());
    }

    return TileAssets(
      id: json['id'] as String,
      name: json['name'] as String,
      model3d: model3d,
      texture: assets?['texture'] as String?,
      icon: assets?['icon'] as String?,
      fallbackColor: _parseColor(fallback?['color'] as String? ?? '#FFFFFF'),
    );
  }

  static Color _parseColor(String hexColor) {
    final hex = hexColor.replaceAll('#', '');
    return Color(int.parse('FF$hex', radix: 16));
  }
}
